pv211_utils/evaluation_metrics.py

```diff
@@ -0,0 +1,405 @@
+"""The evaluation_metrics module provides functions for evaluation of IR systems.
+
+Functions:
+---------
+mean_average_precision(IRSystemBase, OrderedDict, Set[JudgementBase], int, int) -> float:
+    Calculate mean average precision of a system.
+mean_precision(IRSystemBase, OrderedDict, Set[JudgementBase], int, int) -> float:
+    Calculate mean precision of a system.
+mean_recall(IRSystemBase, OrderedDict, Set[JudgementBase], int, int) -> float:
+    Calculate mean recall of a system.
+normalized_discounted_cumulative_gain(IRSystemBase, OrderedDict, Set[JudgementBase], int, int) -> float:
+    Calculate normalized discounted cumulative gain of a system.
+bpref(IRSystemBase, OrderedDict, Set[JudgementBase], int, int) -> float:
+    Calculate mean bpref score of a system.
+"""
+from .entities import JudgementBase, QueryBase
+from .irsystem import IRSystemBase
+
+from typing import Set, OrderedDict
+from multiprocessing import Pool, get_context
+from functools import partial
+from math import log2
+import abc
+
+
+def _judgements_obj_to_id(old_judgements: Set[JudgementBase]) -> Set:
+    new_judgements = set()
+    for q, d in old_judgements:
+        new_judgements.add((q.query_id, d.document_id))
+
+    return new_judgements
+
+
+def _calc_recall(system: IRSystemBase, judgements: Set, k: int,
+                 query: QueryBase) -> float:
+    num_relevant = 0
+    num_relevant_topk = 0
+    current_rank = 1
+
+    for document in system.search(query):
+        if (query.query_id, document.document_id) in judgements:
+            num_relevant += 1
+            if current_rank <= k:
+                num_relevant_topk += 1
+        current_rank += 1
+
+    if num_relevant > 0:
+        recall = num_relevant_topk / num_relevant
+    else:
+        recall = 1
+
+    return recall
+
+
+def _calc_precision(system: IRSystemBase, judgements: Set, k: int,
+                    query: QueryBase) -> float:
+    num_relevant = 0
+    precision = 0.0
+    current_rank = 1
+
+    for document in system.search(query):
+        if current_rank > k:
+            break
+        if (query.query_id, document.document_id) in judgements:
+            num_relevant += 1
+        current_rank += 1
+
+    precision = num_relevant / k
+
+    return precision
+
+
+def _calc_average_precision(system: IRSystemBase, judgements: Set, k: int,
+                            query: QueryBase) -> float:
+    num_relevant = 0
+    average_precision = 0.0
+    current_rank = 1
+
+    for document in system.search(query):
+        if current_rank > k:
+            break
+        if (query.query_id, document.document_id) in judgements:
+            num_relevant += 1
+            average_precision += num_relevant / current_rank
+        current_rank += 1
+
+    average_precision /= num_relevant if num_relevant > 0 else 1
+
+    return average_precision
+
+
+class calc_map(abc.ABC):
```

Comment:
What's the advantage of `evaluation_metrics.calc_map` compared to `evaluation_metrics.mean_average_precision`? Can we remove one in favor of the other?